akka-http-core/src/main/scala/akka/http/engine/server/HttpServerBluePrint.scala

```diff
@@ -240,4 +266,75 @@ private[http] object HttpServerBluePrint {
         case _ ⇒ ctx.fail(error)
       }
   }
+
+  case class WebsocketSetup(
+    flow: Flow[ByteString, ByteString, Any],
+    publisherKey: StreamUtils.ReadableCell[Publisher[FramePart]],
+    subscriberKey: StreamUtils.ReadableCell[Subscriber[FramePart]],
+    switchSource: Source[SwitchToWebsocketToken.type, Any]) extends WebsocketSwitch {
+    @volatile var switchToWebsocketRef: Option[ActorRef] = None
+
+    def switchToWebsocket(handlerFlow: Flow[FramePart, FramePart, Any])(implicit mat: FlowMaterializer): Unit = {
+      // 1. fill processing hole in the websocket pipeline with user-provided handler
+      Source(publisherKey.value)
+        .via(handlerFlow)
+        .to(Sink(subscriberKey.value))
+        .run()
+
+      // FIXME: is there a race condition between 1 and 2?
+
+      // 2. flip the switch
+      switchToWebsocketRef.get ! TokenSourceActor.Trigger
+    }
+  }
+  def websocketPipeline: WebsocketSetup = {
+    val sinkCell = new StreamUtils.OneTimeWriteCell[Publisher[FramePart]]
+    val sourceCell = new StreamUtils.OneTimeWriteCell[Subscriber[FramePart]]
+
+    val sink = StreamUtils.oneTimePublisherSink[FramePart](sinkCell, "frameHandler.in")
+    val source = StreamUtils.oneTimeSubscriberSource[FramePart](sourceCell, "frameHandler.out")
+
+    val flow =
+      Flow[ByteString]
+        .transform[FramePart](() ⇒ new FrameEventParser)
+        .via(StreamUtils.sinkSourceFlow(sink, source))
+        .transform(() ⇒ new FrameEventRenderer)
+
+    lazy val setup = WebsocketSetup(flow, sinkCell, sourceCell, switchToWebsocketSource)
+    lazy val switchToWebsocketSource: Source[SwitchToWebsocketToken.type, ActorRef] =
+      Source.actorPublisher[SwitchToWebsocketToken.type] {
+        Props {
+          val actor = new TokenSourceActor(SwitchToWebsocketToken)
+          setup.switchToWebsocketRef = Some(actor.context.self)
+          actor
+        }
```

Comment:
Does this work with `akka.actor.serialize-creators=on`, otherwise move to `props` factory method in `TokenSourceActor` companion.

That usage of `actor.context.self` is "interesting". @rkuhn Is it safe before the actor is started?